test/types/community/collection/findX.test-d.ts

```diff
@@ -300,3 +300,223 @@ expectAssignable<SchemaWithUserDefinedId | null>(await schemaWithUserDefinedId.f
+      }
+    )
+  ).value
+);
+
+// projections do not change the return type - our typing doesn't support this
+expectType<WithId<{ a: number; b: string }> | null>(
+  (await coll.findOneAndDelete({ a: 3 }, { projection: { _id: 0 } })).value
+);
+
+/**
+ * mutually recursive types are not supported and will not get type safety
+ */
+interface A {
+  b: B;
+}
+
+interface B {
+  a: A;
+}
+
+declare const mutuallyRecursive: Collection<A>;
+//@ts-expect-error
+mutuallyRecursive.find({});
+mutuallyRecursive.find({
+  b: {}
+});
+
+/**
+ * types that are not recursive in name but are recursive in structure are
+ *   still supported
+ */
+interface RecursiveButNotReally {
+  a: { a: number; b: string };
+  b: string;
+}
+
+declare const recursiveButNotReallyColl: Collection<RecursiveButNotReally>;
+expectError(
+  recursiveButNotReallyColl.find({
+    'a.a': 'asdf'
+  })
+);
+recursiveButNotReallyColl.find({
+  'a.a': 2
+});
+
+/**
+ * recursive schemas are now supported, but with limited type checking support
+ */
+interface RecursiveSchema {
+  name: RecursiveSchema;
+  age: number;
+}
+
+declare const collection3: Collection<RecursiveSchema>;
+collection3.find({
+  name: {
+    name: {
+      age: 23
+    }
+  }
+});
+
+collection3.find({
+  age: 23
+});
+
+/**
+ * Recursive optional schemas are also supported with the same capabilities as
+ *   standard recursive schemas
+ */
+interface RecursiveOptionalSchema {
+  name?: RecursiveOptionalSchema;
+  age: number;
+}
+
+declare const collection4: Collection<RecursiveOptionalSchema>;
+
+collection4.find({
+  name: {
+    name: {
+      age: 23
+    }
+  }
+});
+
+collection4.find({
+  age: 23
+});
+
+/**
+ * recursive union types are not supported
+ */
+interface Node {
+  next: Node | null;
+}
+
+declare const nodeCollection: Collection<Node>;
+
+// circular type error is thrown
+// @ts-expect-error
+nodeCollection.find({
+  next: null
+});
+
+nodeCollection.find({
+  next: 'asdf'
+});
+
+nodeCollection.find({
+  'next.next': 'asdf'
+});
+
+nodeCollection.find({ 'next.next.next': 'yoohoo' });
+
+/**
+ * Recursive schemas with arrays are also supported
+ */
+interface MongoStrings {
+  projectId: number;
+  branches: Branch[];
+  twoLevelsDeep: {
+    name: string;
+  };
+}
+
+interface Branch {
+  id: number;
+  name: string;
+  title?: string;
+  directories: Directory[];
+}
+
+interface Directory {
+  id: number;
+  name: string;
+  title?: string;
+  branchId: number;
+  files: Directory[];
+}
+
+declare const coll5: Collection<MongoStrings>;
+expectError(
+  coll5.findOne({
+    'branches.0.id': 'hello'
+  })
+);
+
+expectError(
+  coll5.findOne({
+    'branches.0.directories.0.id': 'hello'
+  })
+);
+
+// type safety breaks after the first
+//   level of nested types
+coll5.findOne({
+  'branches.0.directories.0.files.0.id': 'hello'
+});
+
+coll5.findOne({
+  branches: [
+    {
+      id: 'asdf'
+    }
+  ]
+});
+
+// type inference works on non-recursive properties but only at the top level
```

Comment:
I'll fix this comment.  Type inference works on properties whose values are literals and not objects (this is the same as functionality on `main`).